Vector.c

```diff
@@ -93,91 +101,168 @@ void Vector_prune(Vector* this) {
    memset(this->array, '\0', this->arraySize * sizeof(Object*));
 }
 
-//static int comparisons = 0;
-
-static void swap(Object** array, int indexA, int indexB) {
-   assert(indexA >= 0);
-   assert(indexB >= 0);
-   Object* tmp = array[indexA];
-   array[indexA] = array[indexB];
-   array[indexB] = tmp;
+ATTR_NONNULL
+static void swapByte(char* p1, char* p2) {
+    char temp = *p1;
+    *p1 = *p2;
+    *p2 = temp;
 }
 
-static int partition(Object** array, int left, int right, int pivotIndex, Object_Compare compare) {
-   const Object* pivotValue = array[pivotIndex];
-   swap(array, pivotIndex, right);
-   int storeIndex = left;
-   for (int i = left; i < right; i++) {
-      //comparisons++;
-      if (compare(array[i], pivotValue) <= 0) {
-         swap(array, i, storeIndex);
-         storeIndex++;
+ATTR_NONNULL
+static void rotate(void* buffer, size_t leftSize, size_t rightSize) {
```

Comment:
A note to leave to myself:
There are two possible interfaces for this `rotate` function. To minimize the chances that the callers pass  invalid values, I picked the sizes as the second and third parameters. When it comes to optimization, this would be more dependent on the compiler's capability in simplifying additions and subtraction than the version that passes `mid` and `end` pointers. I didn't check the code sizes produced yet, so this remains a TODO for me.

```c
rotate_a(void* buffer, size_t leftSize, size_t rightSize);
rotate_b(void* start, void* mid, const void* end);
```